src/index.js

```diff
@@ -133,47 +133,45 @@ module.exports = class RpcEngine extends SafeEventEmitter {
 
     // go down stack of middleware, call and collect optional returnHandlers
     for (const middleware of this._middleware) {
+      isComplete = await this._runMiddleware(
+        req, res, middleware, returnHandlers,
+      )
       if (isComplete) {
         break
       }
-      await runMiddleware(middleware)
     }
-
     return { isComplete, returnHandlers: returnHandlers.reverse() }
+  }
 
-    // runs an individual middleware
-    function runMiddleware (middleware) {
-      return new Promise((resolve) => {
+  // runs an individual middleware
+  _runMiddleware (req, res, middleware, returnHandlers) {
```

Comment:
Could this be made `static` as well? As a slight optimization.
```suggestion
  static _runMiddleware (req, res, middleware, returnHandlers) {
```

Reply:
Done in: b928e20

Had to change the single `this._runMiddleware` into `RpcEngine._runMiddleware`.